src/nvim/log.c

```diff
@@ -137,6 +163,94 @@ bool logmsg(int log_level, const char *context, const char *func_name,
   return ret;
 }
 
+/// Log an array of lines to $NVIM_LOG_FILE. The only failure condition is that
+/// the log file can't be opened. Any other errors are written to the log file.
+bool do_log_array(char *log_level, Array lines, Dictionary opt)
```

Comment:
The 3 options I considered were:

**A)** Create an array of `char *` with pointers into the VimL Array and pass that along to a generic array logging function in `vim.c`.

This is nice in that `log.c` doesn't need to know anything about VimL types, but then you have to send the error messages through in a 2nd call to the generic logging function, which introduces a race condition where another process may write to the log file in between any error lines and log lines. You can solve that by doing the locking/unlocking of the log file inside the `nvim_log()` API function and doing two consecutive calls to the generic logging function, but now you have a generic log function which relies on the caller to do the locking/unlocking, which feels untidy/unsafe.

**B)** Accept that we're going to have to lock/unlock the inside the `nvim_log()` API function, but instead of allocating potentially huge `char *` arrays up front, just iterate over the VimL array and call a generic line logging function once for each line.

I like that this solution avoids allocating the entire `char *` array up front, but don't like that it results in an extra function call for every single log line. This would mean regenerating the timestamp for every single line, or else you have to outsource *that* to the `nvim_log()` API function as well.

**C)** Just make a function in `log.c` that takes a VimL array.

This is a bit untidy in that `log.c` needs to start understanding VimL types, but it means that the locking, unlocking and timestamp generation stay within `log.c`, _and_ you don't have to allocate the `char *` array at all, _and_ the execution overhead of each log line is just the call to `fprintf()`.

---

The ability to handle unexpectedly huge data sets efficiently has always been one of the big selling points of Vim, so as much as possible I wanted this `nvim_log()` method to also cope well with unexpected huge log arrays, even though it's unlikely any client will ever send through anything more than maybe a 1000-line stack trace. That's why I picked option **C** above.

Thoughts?